### PR TITLE
add tif/asc read and write support

### PR DESCRIPTION
**Pull request by vlandau**
Added two functions to support read and write of tifs and asciis.

### CODE REVIEW

**Comment by vlandau** [comment]
> Ideally we should have only one set of utility functions to read/write. Does Omniscape just use these from CS or should we refactor some of this into a separate package that both packages can then use?

Agreed. I'd like to use them from CS once they are incorporated. I just wanted to add tif support immediately to Omniscape so I hrd coded the functions into that package, knowing that CS might take a bit longer to get them fully built in.

**Comment by vlandau** [comment]
@glaroc might make sense to work from this PR from now on? We can probably give you permissions for the vl/tifs branch.

**Comment by glaroc** [comment]
@vlandau, since this PR only has the two read and write functions in it, it would be easier for me to work from the other PR since there are many more changes there.

**Comment by vlandau** [comment]
Okay, I'll push my functions to your PR for now.

**Comment by glaroc** [comment]
@vlandau Actually I just had a look and since your functions don't produce the same kind of output as the previous functions in Circuitscape, this will require more in-depth changes to Circuitscape.

**Comment by vlandau** [comment]
I'm going to let @ranjanan chime in. They do provide an identical output to the original `read_ascii` function -- an `Array{Float64,2}`. They just also provide additional outputs: the wkt and the transform.

**Comment by glaroc** [comment]
Ok, let me see if I can easily combine the two PRs

**Comment by vlandau** [comment]
Just FWIW, I did not mean to step on toes with my versions of these functions. I wrote them for Omniscape -- I just figured they might also be useful for Circtuiscape :)

**Comment by vlandau** [comment]
I'm happy to push my functions to your PR. I think it's best to wait for guidance from either @ranjanan or @ViralBShah though.

**Comment by ViralBShah** [comment]
@ranjanan and I spoke a little bit yesterday and he will be chiming in.

**Comment by glaroc** [comment]
@vlandau I think it's probably a good idea to use your functions instead of the Geoarrays dep. However, the RasterMeta struct containing the metadata is used in a lot of places in Circuitscape, so it might need to be recreated from your functions.

**Comment by ViralBShah** [comment]
So what needs to be done to get this merged?

**Comment by glaroc** [comment]
In summary, the similar PR I submitted has a GeoArrays dependency, but creates the same metadata (RasterMeta struct) that is currently used in Circuitscape. This PR by @vlandau removes the GeoArrays dep by using ArchGDAL directly, but more work would be needed to handle and trasnfer the raster metadata and integrate it in Circuitscape.

**Comment by vlandau** [comment]
> So what needs to be done to get this merged?

I can work with guidance from @ranjanan to make a pass at incorporating this fully this week. I just copied over the functions for now.

**Comment by vlandau** [comment]
To elaborate, there are a few specific things to complete this PR:
1. Create an .ini argument for writing to tif (should this be the default? it is more commonly used in the GIS world than ASC, but it might throw off some long-time users)
2. Save projection info to RasterMeta to carry throughout.
3. Recompute some of the RasterMeta components (e.g. nrow, ncol) that were originally read with readline/readdlm from the ASC header.

**Comment by ranjanan** [comment]
@vlandau and I spoke and we have a plan for this PR.

**Comment by ranjanan** [comment]
Should get to passing tests today/tomorrow and then be merged. These IO utils will remain in Circuitscape for now, but in the long term be moved out to its own package.

**Comment by vlandau** [comment]
@ranjanan, @ViralBShah some tests are failing, but these are the same tests that are already failing on master, so I think this might end up being a pretty seamless merge! :crossed_fingers: 

There currently are not tests for writing as tif -- though it is almost guaranteed to work because of the nature of the `write_raster` function I developed. It would be worth making sure by setting `write_as_tif = true` in one or more of the test .ini files, but I didn't want to mess anything up with the saved tests outputs.

At this stage this PR should be considered a complete/near-final draft pending addition of any updates to tests to explicitly write as tif. Also presumably the failing tests need to be fixed on master.

**Comment by vlandau** [comment]
Also I'm not sure if the GZip reader is tested explicitly. It may not work with this PR as-is, so we should add a test for it before going ahead.

File: src/io.jl [hunk 1]
```diff
@@ -85,32 +88,51 @@ function read_cellmap(habitat_file::String, is_res::Bool, ::Type{T}) where {T}
...
+            end
         end
-    end
-    map!(x -> x == rastermeta.nodata ? -9999. : x , c, c)
-    close(f)
-    c, rastermeta
+        map!(x -> x == rastermeta.nodata ? -9999. : x , c, c)
+        close(f)
+        return c, rastermeta
+    else=#
```
Comment (vlandau) [inline src/io.jl:115]: This block can be deleted once okay'ed -- we will still need tests and potentially a solution for the Gzip stuff though if we want to keep it.

File: src/io.jl [hunk 2]
```diff
@@ -85,32 +88,51 @@ function read_cellmap(habitat_file::String, is_res::Bool, ::Type{T}) where {T}
...
+    ncols = dims[2]
+    nrows = dims[1]
+    xllcorner = transform[1]
+    yllcorner = nrows - transform[4]
+    cellsize = transform[2]
+    nodata = -9999 # set in read_raster (overwrites old nodata val)
+    RasterMeta(ncols, nrows, xllcorner, yllcorner, cellsize, nodata, transform, wkt)
 end
 
-function _ascii_grid_read_header(habitat_file, f)
```
Comment (vlandau) [inline src/io.jl:113]: This function should be deleted, but kept for now pending any need for it for GZip stuff.

File: src/io.jl (hunk 1 above)
Comment (vlandau) [reply src/io.jl]: deleted it. We'll need a different solution for GZip that was there originally anyway.

**Comment by vlandau** [comment]
closes #220

File: src/out.jl [hunk 3]
```diff
@@ -482,7 +456,56 @@ function write_cum_maps(cum, cellmap::Matrix{T}, cfg, hbmeta, write_max, write_c
...
         postprocess_cum_curmap!(max_curr)
-        write_aagrid(max_curr, "", cfg, hbmeta, max = true)
+        write_grid(max_curr, "", cfg, hbmeta, max = true)
+    end
+
+end
+
+# Write a single band raster, either in .tif or .asc format,
+# inspired by GeoArrays.write()
+function write_raster(fn_prefix::AbstractString,
```
Comment (ranjanan) [inline src/out.jl:466]: Why is this an `AbstractString` as opposed to a `String`

File: src/out.jl [hunk 4]
```diff
@@ -482,7 +456,56 @@ function write_cum_maps(cum, cellmap::Matrix{T}, cfg, hbmeta, write_max, write_c
...
+                      array,
+                      wkt::AbstractString,
```
Comment (ranjanan) [inline src/out.jl:468]: This too

File: src/out.jl (hunk 3 above)
Comment (vlandau) [reply src/out.jl:466]: Should be able to safely change. I just wanted to err on the side of more general :slightly_smiling_face:

File: src/out.jl (hunk 4 above)
Comment (vlandau) [reply src/out.jl:468]: Adding a type for `array` would be ideal here, but I was having trouble (:facepalm:) figuring out how to make it accept any numeric array -- kept getting errors about no matching method no matter what I tried.

**Comment by glaroc** [comment]
I would be nice to also have possibility of having pairs as .tif raster as well. 
https://github.com/Circuitscape/Circuitscape.jl/blob/b6432139e53fb63d283113d1b07ee0b163dc90ac/src/raster/pairwise.jl#L15

**Comment by vlandau** [comment]
@glaroc do you mean the source node raster? This PR should support reading in tifs of source nodes. Basically any file that was an ASCII raster can now also be a tif.

**Comment by glaroc** [comment]
@vlandau Yes. It is not working for me if the source and destination file is a raster. I feel further modifcations are needed here https://github.com/Circuitscape/Circuitscape.jl/blob/b6432139e53fb63d283113d1b07ee0b163dc90ac/src/io.jl#L211

**Comment by vlandau** [comment]
Ah I see. Are you able use an ASCII raster of ground points before? Looks like it was just a .txt file?

**Comment by glaroc** [comment]
no, it was possile to use an .asc file with zones of pixels for source and destinations. See my PR if you want inspiration for changes needed: 

https://github.com/glaroc/Circuitscape.jl/blob/c8a3ed7607e0af8d9e896bb96016b66a51601d7b/src/io.jl#L315

**Comment by vlandau** [comment]
Sorry I mean was it an ASCII raster with a header?

**Comment by glaroc** [comment]
@vlandau yes

**Comment by vlandau** [comment]
Cool I'll look into it. Thanks!

**Comment by vlandau** [comment]
Okay I see the problem. Thanks for pointing that out.

Just fixed it and I'm testing it now. Will push once I verify it's working.

**Comment by vlandau** [comment]
Okay, just pushed, let me know if the issue isn't resolved.

**Comment by glaroc** [comment]
@vlandau It works! Thanks!

**Comment by ranjanan** [comment]
@vlandau on the following ASCII file: 
```
ncols         5
nrows         5
xllcorner     100
yllcorner     100
cellsize      10
NODATA_value  -9999
1       1       -9999   1       1
1       -9999   -9999   -9999   -9999
1       -9999   -9999   1       1
1       -9999   -9999   1       1
1       -9999   -9999   1       1
```
`_grid_reader` returns -145 instead of 100. This is what is causing some of tests to error out. Why are we calculating `yllcorner` there? I thought it came from the file. 

Is there an easy fix?

**Comment by vlandau** [comment]
Hmm, let me look into it. I thought the same test was failing on master as well.

**Comment by vlandau** [comment]
Okay I think I might know what's going on. Will post back here shortly.

**Comment by ranjanan** [comment]
Yes, but why aren't we directly "reading it off" the file? That's what used to happen in the old code.

**Comment by vlandau** [comment]
> Yes, but why aren't we directly "reading it off" the file? That's what used to happen in the old code. 

Sorry just edited my previous comment as you posted this. Misunderstood your question at first.

**Comment by vlandau** [comment]
Down to just one test failing... sgVerify15, I can look into it. It's a GDAL-related error.
```julia
[ Info: Testing sgVerify15
[ Info: 2020-05-14 16:33:34 : Precision used: Double
[ Info: 2020-05-14 16:33:34 : Reading maps
Raster Pairwise: Error During Test at /home/travis/build/Circuitscape/Circuitscape.jl/src/utils.jl:291
  Got exception outside of a @test
  GDALError (CE_Failure, code 10):
  	Pointer 'hDS' is NULL in 'GDALGetGeoTransform'.
  
  
  Stacktrace:
   [1] gdaljl_errorhandler(::GDAL.CPLErr, ::Int32, ::Cstring) at /home/travis/.julia/packages/GDAL/XGYqJ/src/error.jl:36
   [2] gdalgetgeotransform at /home/travis/.julia/packages/GDAL/XGYqJ/src/gdal_h.jl:1109 [inlined]
   [3] getgeotransform! at /home/travis/.julia/packages/ArchGDAL/FiLcr/src/dataset.jl:638 [inlined]
   [4] getgeotransform(::ArchGDAL.Dataset) at /home/travis/.julia/packages/ArchGDAL/FiLcr/src/dataset.jl:643
   [5] read_raster(::String, ::Type) at /home/travis/build/Circuitscape/Circuitscape.jl/src/io.jl:440
   [6] _grid_reader(::Type, ::String) at /home/travis/build/Circuitscape/Circuitscape.jl/src/io.jl:91
   [7] read_cellmap(::String, ::Bool, ::Type{Float64}) at /home/travis/build/Circuitscape/Circuitscape.jl/src/io.jl:70
   [8] load_raster_data(::Type, ::Type, ::Dict{String,String}) at /home/travis/build/Circuitscape/Circuitscape.jl/src/io.jl:380
   [9] raster_pairwise(::Type, ::Type, ::Dict{String,String}) at /home/travis/build/Circuitscape/Circuitscape.jl/src/raster/pairwise.jl:18
   [10] _compute(::Type, ::Type, ::Dict{String,String}) at /home/travis/build/Circuitscape/Circuitscape.jl/src/run.jl:42
   [11] macro expansion at ./util.jl:212 [inlined]
   [12] compute(::String) at /home/travis/build/Circuitscape/Circuitscape.jl/src/run.jl:31
   [13] runtests(::typeof(compute)) at /home/travis/build/Circuitscape/Circuitscape.jl/src/utils.jl:299
   [14] top-level scope at /home/travis/build/Circuitscape/Circuitscape.jl/test/runtests.jl:12
   [15] include at ./boot.jl:328 [inlined]
   [16] include_relative(::Module, ::String) at ./loading.jl:1105
   [17] include(::Module, ::String) at ./Base.jl:31
   [18] include(::String) at ./client.jl:424
   [19] top-level scope at none:6
   [20] eval(::Module, ::Any) at ./boot.jl:330
   [21] exec_options(::Base.JLOptions) at ./client.jl:263
   [22] _start() at ./client.jl:460
```

**Comment by vlandau** [comment]
Ah, that's the .gz one! So, we need to make a decision about whether we want to continue supporting .gz (but _only_ for ASCII rasters), or just drop it since .tif's are in a compressed format and that solves the issue of wanting to get smaller input file sizes.

**Comment by vlandau** [comment]
I'll add that it would require using the old version of `_ascii_grid_reader`, which is very inefficient and memory-intensive for large files anyway, so that's another point for dropping .gz support IMO.

**Comment by ranjanan** [comment]
Right. I don't mine dropping it. Could we add an error message telling people we don't support .gz files anymore and they should use tifs instead?

**Comment by vlandau** [comment]
Yeah that's a good idea. I can add that in `_grid_reader`. 

Is a normal `@error` fine?

**Comment by vlandau** [comment]
I guess this would be a breaking change. What's sort of release do you anticipate for this. Are you thinking a major release after multithreading is also done?

**Comment by vlandau** [comment]
Just figured out how to read .gz files with ArchGDAL! Had to dig into the GDAL documentation to figure it out. Will make a commit here shortly.

**Comment by ranjanan** [comment]
> Is a normal @error fine? 

Yep, should be good.

**Comment by vlandau** [comment]
One test is still failing, but only on Mac. Everything is passing on Linux. I'm not sure which test it is or what is going on. [The stracktrace](https://travis-ci.org/github/Circuitscape/Circuitscape.jl/jobs/687200993#L2883-L2895) in Travis isn't particularly helpful to a novice like me.

**Comment by vlandau** [comment]
Dumping this here so it's not lost -- here's the stack trace from the failing test on mac os.
```Julia
Raster Pairwise: Error During Test at /Users/travis/build/Circuitscape/Circuitscape.jl/src/utils.jl:291
  Got exception outside of a @test
  GDALError (CE_Failure, code 4):
  	Unable to create file output/sgVerify11_curmap_4_7.asc.
  
  Stacktrace:
   [1] (::Base.var"#732#734")(::Task) at ./asyncmap.jl:178
   [2] foreach(::Base.var"#732#734", ::Array{Any,1}) at ./abstractarray.jl:1920
   [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::UnitRange{Int64}) at ./asyncmap.jl:178
   [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Distributed.var"#208#211"{Distributed.WorkerPool}, ::Function, ::UnitRange{Int64}) at ./asyncmap.jl:154
   [5] #async_usemap#717(::Function, ::Nothing, ::typeof(Base.async_usemap), ::Circuitscape.var"#89#94"{Circuitscape.var"#f#91"{Float64,Int64,Circuitscape.RasterFlags,Dict{String,String},Bool,Array{Int64,1},Array{Tuple{Int64,Int64},1},Array{Int64,1},Circuitscape.Cumulative{Float64},Circuitscape.Shortcut{Float64},Array{Int64,1},Array{Int64,1},SparseArrays.SparseMatrixCSC{Float64,Int64},AlgebraicMultigrid.Preconditioner{AlgebraicMultigrid.MultiLevel{AlgebraicMultigrid.Pinv{Float64},AlgebraicMultigrid.GaussSeidel{AlgebraicMultigrid.SymmetricSweep},AlgebraicMultigrid.GaussSeidel{AlgebraicMultigrid.SymmetricSweep},SparseArrays.SparseMatrixCSC{Float64,Int64},SparseArrays.SparseMatrixCSC{Float64,Int64},LinearAlgebra.Adjoint{Float64,SparseArrays.SparseMatrixCSC{Float64,Int64}},AlgebraicMultigrid.MultiLevelWorkspace{Array{Float64,1},1}}},Circuitscape.ComponentData{Float64,Int64}}}, ::UnitRange{Int64}) at ./asyncmap.jl:103
   [6] (::Base.var"#kw##async_usemap")(::NamedTuple{(:ntasks, :batch_size),Tuple{Distributed.var"#208#211"{Distributed.WorkerPool},Nothing}}, ::typeof(Base.async_usemap), ::Function, ::UnitRange{Int64}) at ./none:0
   [7] #asyncmap#716 at ./asyncmap.jl:81 [inlined]
   [8] #asyncmap at ./none:0 [inlined]
   [9] #pmap#207(::Bool, ::Int64, ::Nothing, ::Array{Any,1}, ::Nothing, ::typeof(Distributed.pmap), ::Function, ::Distributed.WorkerPool, ::UnitRange{Int64}) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:126
   [10] pmap(::Function, ::Distributed.WorkerPool, ::UnitRange{Int64}) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:101
   [11] #pmap#217(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Distributed.pmap), ::Function, ::UnitRange{Int64}) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156
   [12] pmap at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Distributed/src/pmap.jl:156 [inlined]
   [13] amg_solver_path(::Circuitscape.GraphData{Float64,Int64}, ::Circuitscape.RasterFlags, ::Dict{String,String}, ::Bool) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/core.jl:221
   [14] single_ground_all_pairs(::Circuitscape.GraphData{Float64,Int64}, ::Circuitscape.RasterFlags, ::Dict{String,String}, ::Bool) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/core.jl:55
   [15] _pt_file_polygons_path(::Circuitscape.RasData{Float64,Int64}, ::Circuitscape.RasterFlags, ::Dict{String,String}) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/raster/pairwise.jl:105
   [16] raster_pairwise(::Type, ::Type, ::Dict{String,String}) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/raster/pairwise.jl:27
   [17] _compute(::Type, ::Type, ::Dict{String,String}) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/run.jl:42
   [18] macro expansion at ./util.jl:212 [inlined]
   [19] compute(::String) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/run.jl:31
   [20] runtests(::typeof(compute)) at /Users/travis/build/Circuitscape/Circuitscape.jl/src/utils.jl:299
   [21] top-level scope at /Users/travis/build/Circuitscape/Circuitscape.jl/test/runtests.jl:12
   [22] include at ./boot.jl:328 [inlined]
   [23] include_relative(::Module, ::String) at ./loading.jl:1105
   [24] include(::Module, ::String) at ./Base.jl:31
   [25] include(::String) at ./client.jl:424
   [26] top-level scope at none:6
   [27] eval(::Module, ::Any) at ./boot.jl:330
   [28] exec_options(::Base.JLOptions) at ./client.jl:263
   [29] _start() at ./client.jl:460
```

**Comment by vlandau** [comment]
Yay... [new error](https://travis-ci.org/github/Circuitscape/Circuitscape.jl/jobs/687848514#L5368) on Mac OS -- we're peeling the onion slowly but surely haha

**Comment by vlandau** [comment]
In theory the below (run from within the Circuitscape repo) should reproduce the error on a Mac. It works just fine on my Linux system, though.

```julia
using GZip
include("src/consts.jl")

filename = "test/input/raster/all_to_one/12/include_matrix.txt"

f = endswith(filename, "gz") ? GZip.open(filename, "r") : open(filename, "r")
filetype = _guess_file_type(filename, f)
close(f)

open(filename, "r") do f
    minval = parse(Float64, split(readline(f))[2])
end
```

**Comment by ranjanan** [comment]
Ran that code on my mac and didn't get an error. The VM's are fairly low end, which might be why these tests are failing. You can put this in the `runtests.jl` and check if it errors out on Mac.

**Comment by ranjanan** [comment]
[Found the error GDAL error description here](https://gdal.org/api/cpl.html?highlight=ce_failure#_CPPv48CPLError6CPLErr11CPLErrorNumPKcz). Unfortunately, it isn't very informative. Code 4 simply means File IO error.

**Comment by vlandau** [comment]
We're back to too many open files.

I've checked every instance of `open()` in the entire codebase, and they all either used a `do` block or have a corresponding `close()`. I'm not sure what else to look for. There's [an `open()`](https://github.com/Circuitscape/Circuitscape.jl/blob/vl/tifs/src/logging.jl#L46) related to logging, but based on the Coveralls report it looks like that line is never hit anyway.

**Comment by vlandau** [comment]
See [julia/issues/8891](https://github.com/JuliaLang/julia/issues/8891). Possibly related? It's a really old issue though, but it hasn't been closed.

**Comment by vlandau** [comment]
Okay, tests are all passing now after adding a manual `GC.gc()` after each test run. This does not solve the problem we had with Circuitscape leaving files opened -- instead it's just a hacky fix. Because of that, I'm not sure we should merge this PR just yet.

**Comment by vlandau** [comment]
Hooray! We've got tests passing on Mac and Linux. The test failing on Appveyor is the same one failing on master. I think we are finally ready to merge @ranjanan! :tada: 

Though note that whatever issue occurring on Windows seemed to be resolved by running GC.gc() after each test... not sure what's up with that. I removed GC because in theory if we are bug free we shouldn't need to be running that after tests (didn't want to be inadvertently hiding bugs).

cc @ViralBShah and @glaroc

**Pull request merged**